### ISSUE TITLE
Is there a synchronous way to see if the user is authenticated?

### ISSUE THREAD

**Issue by jrp-23** [aws-amplify/amplify-js#1792]
For my SPA I want to check if the user is authenticated prior to navigating to a route.  I understand that CurrentAuthenticatedUser is used for this, but it is promise-based and I have to show a loading gif each time the user navigates (to wait for the promise to resolve).

Is there a synchronous way to check if the user is authenticated so that I don’t have to show the loading gif each time the user navigates?

**Comment by haverchuck** [comment]
@jrp-23  - We do not have this at the moment, but I am marking this issue as a feature request.

**Comment by borisdayma** [comment]
I'm curious to know what is the best way to handle it at the moment. I need to check if the user is authenticated prior to opening a page and redirect otherwise and the promise makes it difficult.

**Comment by codeandsugar** [comment]
> I'm curious to know what is the best way to handle it at the moment. I need to check if the user is authenticated prior to opening a page and redirect otherwise and the promise makes it difficult.

Any progress on this?

I've been struggling with this as well.  My application is routing to the Cognito hosted UI for authentication, and it takes 2-3 log in attempts before the Promise gets resolved.  This is is a cumbersome user experience, and I can't find a solid non-trivial coding example for how to do this properly.

I'm using Angular 7.2 with Amplify 1.2.22.

**Comment by bezreyhan** [comment]
Is there any progress on this feature or a recommended workaround?

We need to be able to synchronously check if there is a cached user. It is causing significant issues for us. Thank you for the help.

**Comment by souravs-2pirad** [comment]
Yes, there needs to be a synchronous way to check if a user exists in localstorage.
Waiting for promise to be resolved takes time ===> unwanted.

Any help would be appreciated :)

**Comment by ammanvedi** [comment]
Agree with all the points above 👍

**Comment by jamesaucode** [comment]
Context: `Auth.currentAuthenticatedUser()` is the recommended way to check if a user is authenticated, however, it returns a promise, developers want a 100% synchronous way to check the authentication status.
After doing some research, it is possible to validate the identity if a user is authenticated by looking at localStorage (or any kind of `local storage` defined by the user), and verify if there’s a valid token without invoking a refreshToken call to make it completely synchronous.

The major hurdles are:
1. OAuth flow, afaik the OAuth flow is handled in `Auth.currentAuthenticatedUser()`, and it is not synchronous.
2. Storage sync, we want to keep the function generic enough so React Native users who uses `AsyncStorage` can also use it without any issues (see: #2337), but waiting for storage to sync is asynchronous.
An `Auth.isAuthenticated()` API that doesn’t consider the above 2 scenarios might cause further confusion from our users, therefore we are still researching a proper way to implement this.  Will provide an update on this as we continue our investigation.

**Comment by returnzer0** [comment]
One other way could be to attach listeners, so that if/when the token refreshes or fails to refresh you can trigger some kind of a hook (could be a context API call or some call to a state variable). It will update that variable every so often and the developers can rely on that variable to see synchronously if a user is authenticated or not.

For our case that would work as well.

**Comment by gillkyle** [comment]
I found a workaround that gets the job done for me, I'm using Amplify authentication to connect to Cognito without the rest of Amplify so I have all the Cognito variables I need to configure things as environment variables:

```
AWS_COGNITO_USER_POOL_CLIENT_ID=1astlfkpcme1dfomeockiumd5a
```

Amplify saves the `idToken`, `refreshToken`, and `accessToken` with a key in `localStorage` that includes the user pool client ID _as well as_ the last authenticated user's id (which is also stored in local storage). They look sort of like this:

```
CognitoIdentityServiceProvider.1astlfkpcme1dfomeockiumd5a.5c9d74e0-da8c-4fc4-bf28-243212874836.accessToken
```

The breakdown of that key would be:

```
CognitoIdentityServiceProvider.<user-pool-id>.<last-authed-user-id>.<token-type>
```

Since `localStorage` is synchronous, that means I can make 2 synchronous `localStorage.getItem`s that first grab me the user id, and then follow up with the `accessToken` before I make a request.

```
// synchronously get the accessToken from localStorage since it should be there after authenticating
const lastUserIdFromLocalStorage = localStorage.getItem(
  `CognitoIdentityServiceProvider.${process.env.AWS_COGNITO_USER_POOL_CLIENT_ID}.LastAuthUser`
)
const accessToken = localStorage.getItem(
  `CognitoIdentityServiceProvider.${process.env.AWS_COGNITO_USER_POOL_CLIENT_ID}.${lastUserIdFromLocalStorage}.accessToken`
)

Seems to be working for me.

**Comment by roastdev** [comment]
any update on this? this is super weird that there isn't a synchronous way to get user authentication. I need to toggle routes based on authentication and there's 0 reason to have a loading screen of any sort for this

**Comment by fabilo-au** [comment]
Any further update on this?

**Comment by emmanuelchucks** [comment]
Using a cache like [Tanstack Query](https://github.com/TanStack/query) can help. You might even end up using it for everything :)

**Comment by devdeveloperdev** [comment]
> Using a cache like [Tanstack Query](https://github.com/TanStack/query) can help. You might even end up using it for everything :)

I'm using Tanstack Query as well but unfortunately that doesn't really solve the issue at hand. The reason to use amplify for the client is they abstract a lot of the authentication related tasks such as token refresh, fetch, ect... You can cache it (this is what I'm doing as a workaround using localstorage) but now this is an antipattern. Hopefully they fix this or offer a resolution.
